File: Packages/SecuPersoDomain/Sources/SecuPersoDomain/RiskScoringEngine.swift
import Foundation

public struct RiskScoringEngine: Sendable {
    public init() {}

    public func score(
        exposures: [ExposureRecord],
        logins: [LoginEvent],
        incidents: [IncidentCase],
        now: Date = Date()
    ) -> RiskSnapshot {
        let exposureScore = exposures
            .filter { $0.status == .open }
            .reduce(0) { partialResult, record in
                partialResult + weight(for: record.severity)
            }

        let suspiciousLogins = logins.filter { $0.suspicious || !$0.expected }
        let loginScore = suspiciousLogins.count * 15

        let burstThreshold = now.addingTimeInterval(-(24 * 60 * 60))
        let burstScore = suspiciousLogins.filter { $0.occurredAt >= burstThreshold }.count >= 2 ? 10 : 0

        let unresolvedIncidents = incidents.filter { $0.status == .open }
        let incidentPenalty = unresolvedIncidents.count * 10

        let total = min(100, max(0, exposureScore + loginScore + burstScore + incidentPenalty))
        let level: RiskLevel
        switch total {
        case 70...100:
            level = .high
        case 30...69:
            level = .medium
        default:
            level = .low
        }

        return RiskSnapshot(score: total, level: level, lastUpdatedAt: now)
    }

    private func weight(for severity: ExposureSeverity) -> Int {
        switch severity {
        case .critical:
            return 30
        case .high:
            return 20
        case .medium:
            return 10
        case .low:
            return 5
        }
    }
}

File: src/domain/riskScoringEngine.ts
import type { ExposureRecord, LoginEvent, IncidentCase, RiskSnapshot, RiskLevel, ExposureSeverity } from './models';

const SEVERITY_WEIGHTS: Record<ExposureSeverity, number> = {
  critical: 30,
  high: 20,
  medium: 10,
  low: 5,
};

function severityWeight(severity: ExposureSeverity): number {
  return SEVERITY_WEIGHTS[severity];
}

function classifyRisk(total: number): RiskLevel {
  if (total >= 70) return 'high';
  if (total >= 30) return 'medium';
  return 'low';
}

export interface RiskScoringEngine {
  score(
    exposures: ExposureRecord[],
    logins: LoginEvent[],
    incidents: IncidentCase[],
    now?: Date,
  ): RiskSnapshot;
}

export const riskScoringEngine: RiskScoringEngine = {
  score(exposures, logins, incidents, now = new Date()) {
    const exposureScore = exposures
      .filter((e) => e.status === 'open')
      .reduce((sum, e) => sum + severityWeight(e.severity), 0);

    const suspiciousLogins = logins.filter((l) => l.suspicious || !l.expected);
    const loginScore = suspiciousLogins.length * 15;

    const burstThreshold = new Date(now.getTime() - 24 * 60 * 60 * 1000);
    const burstScore = suspiciousLogins.filter((l) => new Date(l.occurredAt) >= burstThreshold).length >= 2 ? 10 : 0;

    const unresolvedIncidents = incidents.filter((i) => i.status === 'open');
    const incidentPenalty = unresolvedIncidents.length * 10;

    const total = Math.min(100, Math.max(0, exposureScore + loginScore + burstScore + incidentPenalty));

    return {
      score: total,
      level: classifyRisk(total),
      lastUpdatedAt: now.toISOString(),
    };
  },
};